core/src/main/java/dk/alexandra/fresco/framework/network/async/AsyncNetwork.java

```diff
@@ -0,0 +1,336 @@
+        Future<Map<Integer, SocketChannel>> f = cs.poll(timeout.toMillis(), TimeUnit.MILLISECONDS);
+        if (f == null) {
+          throw new TimeoutException("Timed out");
+        } else {
+          this.channelMap.putAll(f.get());
+        }
+      }
+    } catch (InterruptedException e) {
+      close();
+      throw new RuntimeException("Interrupted while connecting network", e);
+    } catch (ExecutionException e) {
+      close();
+      throw new RuntimeException("Failed to connect network", e.getCause());
+    } catch (TimeoutException e) {
+      close();
+      throw new RuntimeException("Timed out connecting network", e);
+    } finally {
+      es.shutdownNow();
+    }
+  }
+
+  private void startReceivers() {
+    for (Entry<Integer, SocketChannel> entry : this.channelMap.entrySet()) {
+      final int id = entry.getKey();
+      SocketChannel channel = entry.getValue();
+      this.receiverService.submit(() -> {
+        try {
+          while (true) {
+            ByteBuffer inBuf = ByteBuffer.allocate(Integer.BYTES);
+            channel.read(inBuf);
+            inBuf.flip();
+            int nextMessageSize = inBuf.getInt();
+            inBuf = ByteBuffer.allocate(nextMessageSize);
+            while (inBuf.remaining() > 0) {
+              channel.read(inBuf);
+            }
+            logger.debug("P{} just received {} bytes from P{}", conf.getMyId(),
+                inBuf.array().length, id);
+            queues.get(id).put(inBuf.array());
+          }
+        } catch (IOException e) {
+          receiveException = e;
+          throw e;
+        } finally {
+          channel.close();
+        }
+      });
+    }
+  }
+
+  /**
+   * Makes connections to the opposing parties.
+   * <p>
+   * The resulting connections will be used to send messages.
+   * </p>
+   *
+   * @throws InterruptedException thrown if interrupted while waiting to do a connection attempt
+   */
+  private Map<Integer, SocketChannel> connectClient() throws InterruptedException {
+    Map<Integer, SocketChannel> channelMap = new HashMap<>(conf.noOfParties() - conf.getMyId());
+    for (int i = conf.getMyId() + 1; i <= conf.noOfParties(); i++) {
+      Party p = conf.getParty(i);
+      SocketAddress addr = new InetSocketAddress(p.getHostname(), p.getPort());
+      boolean connectionMade = false;
+      int attempts = 0;
+      while (!connectionMade) {
+        try {
+          SocketChannel channel = SocketChannel.open();
+          channel.connect(addr);
+          channel.configureBlocking(true);
+          this.channelMap.put(i, channel);
+          ByteBuffer b = ByteBuffer.allocate(1);
+          b.put((byte) conf.getMyId());
+          b.position(0);
+          channel.write(b);
+          connectionMade = true;
+          channelMap.put(i, channel);
+          logger.info("P{} connected to {}", conf.getMyId(), p);
+        } catch (IOException e) {
+          attempts++;
+          Thread.sleep(1 << attempts);
```

Comment:
Create a variable, timeout and call timeout=timeout*timeout. This is unbearable :)

Reply:
Aw, I thought this was pretty elegant really.

Reply:
hmm 
github does not handle comments via mail properly
"It is when bitlogic is close by. In this it is not - so you are forcing
the reader to turn on the bitter. Which for some might be fine, however
to me it seems like a cumbersome way of expressing the obvious"